### PR TITLE
Implement code fix testing in StyleCopTester

### PR DESCRIPTION
**Pull request by pdelvo**
When trying it with roslyn it reports a bunch of [exceptions](https://gist.github.com/pdelvo/d5e7792797b66b5ff8ee) for renames. We should look at. I dont know if its our fault.

### CODE REVIEW

File: StyleCop.Analyzers/StyleCopTester/Program.cs [hunk 1]
```diff
@@ -158,7 +218,31 @@ private static ImmutableArray<DiagnosticAnalyzer> GetAllAnalyzers()
             return analyzers.ToImmutableArray();
         }
 
-        private static async Task<ImmutableList<Diagnostic>> GetAnalyzerDiagnosticsAsync(Solution solution, ImmutableArray<DiagnosticAnalyzer> analyzers)
+        private static ImmutableDictionary<string, ImmutableList<CodeFixProvider>> GetAllCodeFixers()
+        {
+            Assembly assembly = typeof(StyleCop.Analyzers.NoCodeFixAttribute).Assembly;
+
+            var diagnosticAnalyzerType = typeof(CodeFixProvider);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCopTester/Program.cs]: :bulb: Poor variable name choice.

**Comment by sharwell** [comment]
This appears to use incremental fix calculation. While the ability to run these is great to have (for finding exceptions), it's much more important to be able to run fix-all providers, as they are the performance bottleneck when working in large solutions.

File: StyleCop.Analyzers/StyleCopTester/Program.cs [hunk 2]
```diff
@@ -205,6 +289,7 @@ private static void PrintHelp()
             Console.WriteLine("Options:");
             Console.WriteLine("/all     Run all StyleCopAnalyzers analyzers, including ones that are disabled by default");
             Console.WriteLine("/nostats Disable the display of statistics");
+            Console.WriteLine("/nostats Disable the calculation of code fixes");
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCopTester/Program.cs]: :exclamation: Wrong argument text (copy paste bug).

File: StyleCop.Analyzers/StyleCopTester/Extensions.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                dictionary[key] = items.Add(item);
+            }
+            else
+            {
+                dictionary.Add(key, ImmutableList.Create(item));
+            }
+        }
+
+        internal static void AddToInnerList<TKey, TValue>(this IDictionary<TKey, ImmutableHashSet<TValue>> dictionary, TKey key, TValue item)
+        {
```
Comment (otac0n) [inline StyleCop.Analyzers/StyleCopTester/Extensions.cs]: AddToInnerSet?

File: StyleCop.Analyzers/StyleCopTester/Program.cs [hunk 4]
```diff
@@ -205,6 +356,8 @@ private static void PrintHelp()
             Console.WriteLine("Options:");
             Console.WriteLine("/all     Run all StyleCopAnalyzers analyzers, including ones that are disabled by default");
             Console.WriteLine("/nostats Disable the display of statistics");
+            Console.WriteLine("/codefixes Test single code fixes");
+            Console.WriteLine("/fixall Test fix all providers");
```
Comment (otac0n) [inline StyleCop.Analyzers/StyleCopTester/Program.cs]: Horizontal alignment.

File: StyleCop.Analyzers/StyleCopTester/Program.cs [hunk 5]
```diff
@@ -79,7 +81,115 @@ private static void Main(string[] args)
...
+                {
+                    fix.GetOperationsAsync(CancellationToken.None).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    // Report thrown exceptions
+                    lock (lockObject)
+                    {
+                        Console.ForegroundColor = ConsoleColor.Yellow;
+                        Console.WriteLine($"The fix '{fix.Title} 'threw an exception:");
```
Comment (otac0n) [inline StyleCop.Analyzers/StyleCopTester/Program.cs]: Weird spacing around single qoute.

File: StyleCop.Analyzers/StyleCopTester/Program.cs [hunk 6]
```diff
@@ -79,7 +81,115 @@ private static void Main(string[] args)
...
+                        Console.ForegroundColor = ConsoleColor.Red;
+                        Console.WriteLine(ex);
+                        Console.ResetColor();
```
Comment (otac0n) [inline StyleCop.Analyzers/StyleCopTester/Program.cs]: Why manually set the color and call reset color here, yet on lines 126-127 you call the convenience method?

**Comment by pdelvo** [comment]
@otac0n I implemented your feedback

**Comment by otac0n** [comment]
:+1:

**Pull request merged**